File: 1_Two_Sum/solution.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        /*
            Input: vector of integers nums, integer target
            Output: vector of integers containing the index of the two numbers in nums that add up to target

            Constraints
            - cannot use same element twice
            - nums has a minimum length of 2

            Can assume that we have exactly one solution

            Sample test case
            1.) nums = [2,7,11,15], target = 9, output = [0,1]
                2 + 7 = 9, return 0 and 1
            
            Strategy
            1.) Brute force O(n2)
                - For every element, search the remaining elements if they add up to target. If yes return their index.
                bool found = false;
                std::vector<int> res;
                for (int i=0; i<nums.size()-1; i++) {
                    if (found){
                        break;
                    }
                    for (int j=i+1; j<nums.size(); j++) {
                        if (nums[i] + nums[j] == target) {
                            res.push_back(i);
                            res.push_back(j);
                            found = true;
                            break;
                        }
                    }
                }
                return res;
            
            2.) Linear search with hashmap
                - In similar fashion to checking if the other elements add up with the current element to target,
                we can ask if the current element is a complement to the previously found element that will sum to target.
                - In the first example,
                    nums = [2,7,11,15]
                    - In one loop, we ask if target - nums[i] is in the hashmap, if yes, we have the two index.
                    The hashmap will store the complement as key and index as value.

        */
        std::unordered_map<int,int>hashmap;
        std::vector<int> res;
        for (int i=0; i<nums.size(); i++){
            if (hashmap.find(nums[i]) != hashmap.end()) {
                res.push_back(hashmap[nums[i]]);
                res.push_back(i);
                break;
            }else{
                hashmap[target-nums[i]] = i;
            }
        }
        return res;
    }
};

File: 1_Two_Sum/solution.java
import java.util.HashMap;
class Solution {
    public int[] twoSum(int[] nums, int target) {
        /*
            Input: integer array nums, integer target
            Output: integer array containing the two elements in nums summing in target

            - may not use same element twice. i.e. nums = [3,2,4], target = 6; not accepting [3,3]
            - only one answer exists

            Strategy
            1.) hashmap to store nums as complement to downstream element. values are index in nums TC: O(n), SC: O(n)
                - for each element, we find its complement (target-num) in the hashset. If not found, add itself into the map.

                HashMap <Integer, Integer> complements = new HashMap<Integer, Integer>();
                int [] result = new int[2];

                for (int i=0; i<nums.length; i++) {
                    if (complements.containsKey(target - nums[i])) {
                        result[0] = complements.get(target-nums[i]);
                        result[1] = i;
                        break;
                    }
                    complements.put(nums[i],i);
                }
                return result;
            
        */
        HashMap <Integer, Integer> complements = new HashMap<Integer, Integer>();
        int [] result = new int[2];

        for (int i=0; i<nums.length; i++) {
            if (complements.containsKey(target - nums[i])) {
                result[0] = complements.get(target-nums[i]);
                result[1] = i;
                break;
            }
            complements.put(nums[i],i);
        }
        return result;
        
    }
}